egs/sre08/v1/sid/nnet3/xvector/allocate_egs_seg_v1.py

```diff
@@ -0,0 +1,255 @@
+    random.seed(args.seed)
+    #utt2len is a dict: key=uttid, value=num_of_frames
+    utt2len_filename = "{0}/utt2num_frames".format(args.data_dir)
+    utt2len = get_utt2len(utt2len_filename)
+    #utt2feat is a dict: key=uttid, value=extend_filename_of_featue
+    #utt_list is a list cotains all the original uttid
+    utt2feat_filename = "{0}/feats.scp".format(args.data_dir)
+    utt_list, utt2feat = get_utt2feat(utt2feat_filename)
+    #utt2spk is a dict: key=uttid, value=spkid
+    utt2spk_filename = "{0}/utt2spk".format(args.data_dir)
+    utt2spk, spk2utt = get_utt2spk(utt2spk_filename)
+    spks = spk2utt.keys()
+
+    length_types = []
+    # Generate l1 ... ln
+    if args.randomize_chunk_length == "true":
+        length_types = random_chunk_length(args.min_frames_per_chunk, 
+                args.max_frames_per_chunk, args.kinds_of_length)
+    else:
+        length_types = deterministic_chunk_length(args.min_frames_per_chunk,
+                args.max_frames_per_chunk, args.kinds_of_length)
+    
+    # Generate new feats_li.scp in data_li directory
+    # As we hope the total_frames == frames(feats.scp) * num_repeat
+    # So frmaes(feats_li.scp) == (1/kinds_of_length) * frames(feats.scp) * num_repeat
+    # So for each utterance in feats.scp, the number of it in new feats_li.scp
+    # is max{(|utt|/|li|) * num_repeat / kinds_of_length, 1}
+    for i in range(0, args.kinds_of_length):
+        num_err = 0
+        this_length = length_types[i]
+        this_dir_name = '{0}/data_{1}'.format(args.output_dir, this_length)
+        this_feats_filename = "{0}/feats.scp".format(this_dir_name)
+        if not os.path.exists(this_dir_name):
+            os.makedirs(this_dir_name)
+        
+        try:
+            feats_f = open(this_feats_filename, "w")
+        except Exception as e:
+            sys.exit("{0}: error opening file '{1}'. Error was {2}".format(
+                sys.argv[0], this_feats_filename, repr(e)))
+        
+        this_utt2spk_filename = "{0}/utt2spk".format(this_dir_name)
+        try:
+            spk_f = open(this_utt2spk_filename, "w")
+        except Exception as e:
+            sys.exit("{0}: error opening file '{1}'. Error was {2}".format(
+                sys.argv[0], this_utt2spk_filename, repr(e)))
+        
+        for this_utt_id, this_spk_id in utt2spk:
+            this_extend = utt2feat[this_utt_id]
+            this_utt_len = utt2len[this_utt_id]
+            if this_utt_len < this_length:
+                num_err = num_err + 1
+                if num_err > (0.1 * len(utt_list)):
+                    raise Exception("{0} is not a suitable length".format(this_length))
+                continue
+            else:
+                num_segs = max(int(float(this_utt_len) / this_length * 
+                        args.num_repeats / args.kinds_of_length), 1)
+                this_utt_boundary = int(this_utt_len - this_length)
+                num_segs = min(num_segs, this_utt_boundary)
```

Comment:
add log which describes if this_utt_boundary < num_segs.
this_utt_bounday is not appropriate name! change it to num_distinct_segs or num_available_segs.

Reply:
Maybe you miss understand it. It is a extremely special case. 
Assume the length of utterance is 100 frames, so the index of the matrix will be [0, 99].
Assume this_length is 60 frames, so the start_point should be select from range [0, 40) -- [0, this_utt_boundary). Up to num_segs, I divid the range [0, 40) into "num_segs" pieces. 
Assume num_segs=4 so that "this_seg" will be 10.
So we will select 4 start points from [0, 10), [10, 20), [20, 30), [30, 40) seperately. 
The num_segs want to prevent a extremely special case which "this_utt_boundary" < "num_segs". So this utterance only can be divid into "this_utt_boundary" pieces and the start points will be certain, such as 0, 1, 2, ..this_utt_boundary-1.